### DISCUSSION TITLE
Merge/rebase strategy for 3.2

### DISCUSSION THREAD

**Discussion by peteeckel** [auroraresearchlab/netbox-dns#150] [General]
After some additional trouble with PRs already applied to main (but not easily mergable/rebaseable with dev/refactor), I think it's probably best to do the refactoring and then apply all fixes to the refactored code tree. NetBox < 3.1 support is not really a long term option, so we should IMHO concentrate on 3.2 and the documented plugin API.

I suggest working on the netbox32-changes tree for now, and when 3.2 becomes available just swapping "main" and "netbox32-changes", e.g. to "legacy" and "main", so we don't need to jump through all the hoops merging all the stuff to main (trust me, you don't want that - I tried :-)).

Back-porting some of the fixes to 'legacy' might make sense, but given the fact that the old code base actually builds on a totally unsupported foundation I would not put too much effort into it.

I re-organised all pull requests I did so they can be applied to "netbox32-changes", with #139 as the first one All other PRs I did today are based on the refactored code tree, as merging is a pain in the backside otherwise.

Another difficulty is that #146/[NetBox #8888](https://github.com/netbox-community/netbox/issues/8888) currently breaks the CI chain (and fresh installations of NetBox 3.2 beta in general). I hope that will be resolved soon.

What do you think?

**Comment by hatsat32** [comment]
Hi

I agree with you. NetBox < 3.1 support is not a long term option. Maybe we support 3.1 support for a little time but not more. I think we should focus on 3.2

As you say, directly merging netbox32-changes to main is not possible. So your solution (swapping branches) is good.

CI works again :tada:

**Comment by peteeckel** [comment]
According to @jeremystretch, a cut-off has been declared for version 3.2, so no new features should be expected. I'll concentrate on testing now, and maybe do a slight rework of our code based on the new tutorial he published today (I already see some places where our code could be streamlined and made to use the new Plugin API more efficiently).

3.1 support might become tricky sooner or later ... looking at #142, I'm not sure we can fix it at all, as it seems it's somehow related to the NetBox version. I can reproduce the problem with the latest NetBox version 3.1.9, but not with the older versions I tested so far. If it turns out to be a NetBox issue or somehow related to the unsupported NetBox code we have been using before 3.2 (it might even be somewhere down in the JavaScript hell), it could turn out to be an unresolvable issue - and I suspect more of those are hiding somewhere.

> CI works again 🎉

\o/

**Comment by peteeckel** [comment]
I wasn't quite happy with my own suggestion of swapping branches :-)

PR #153 leads to an identical version (with respect to the code base, not the sequence of commits) of netbox32-changes that was rebased to main, so it can be merged in a fast-forward operation. That was actually easier to do than I expected.

The strategy for support of NetBox 3.0/3.1 would then be to tag the last state of the main branch before merging netbox32-changes and creating a release that supports 3.0/3.1 based on that tag, while starting a new release train for the offically support of NetBox >= 3.2.

**Comment by hatsat32** [comment]
Sorry for the late response. I will prepare new release (0.9.0) which supports NetBox 3.0/3.1 in this week. Then we can merge #153 .

**Comment by peteeckel** [comment]
I have another PR in preparation, based on the tutorial by @jeremystretch. 

The main purpose is to align the API we use with some of the new features NetBox 3.2 introduces and which were not already present in Jeremy's initial PR. In addition, I cleaned up imports a bit (there were some leftover imports not required anymore) and ran pylint over it.

I'd like to run some tests with the new code, but so far I think I can push in the next few days.

**Comment by hatsat32** [comment]
Hi. I released version 0.9.0 and created new branch `legacy` based on `main`. If we want to create new release for netbox 3.1/3.0 , we can use this branch and tag new releases 0.9.1 0.9.2 0.9.3 ... But our main focus should be netbox 3.2 :)

I can merge PR #153 if you think this PR is ready.

**Comment by peteeckel** [comment]
Hi Süleyman, I rebased #153 a final time on the new main branch, and it should be OK now.

When I did the rebase I noticed that the tests broke again, and I found that the culprit was a change introduced in NetBox commit `ba1e6e91b9b1427a04a93aa61155aff4cf6318d5`, where the "model_form" parameter in edit views has been renamed to "form". Fixed that, and now the tests are fine again.

I also adjusted the NetBox minimum version to 3.2-beta2. With your new version comparison code in place that should be fine for final release 3.2 and above as well.

So from my part #153 looks fine, and I'll adjust #156 as well shortly (#153 is not a clean version yet, so #156 should be merged soon, too).

**Comment by peteeckel** [comment]
So, now after NetBox 3.2 was finally released I did a final round of adjustments and rolled the result out on my reference customer's systems.

There were a couple of small issues (resulting in PR #162), but in general everything works as desired. And the best of it is that now we are actually using a supported interface that should be much more stable than before (no more version checks in templates etc., code adjustments to stop minor NetBox upgrades from breaking NetBox DNS etc.), so we can move on to implement new features.

Again, a big "thank you" to @jeremystretch for starting the plugin working group!

**Reply by jeremystretch** [reply]
Happy to help, and thank _you_ for your valuable contributions!

Also, congratulations on having the fourth most-popular plugin on our [community survey](https://github.com/netbox-community/netbox/discussions/9026)!
